### PR TITLE
Prometheus endpoint

### PR DESCRIPTION
**Pull request by tonysun83**
## Overview

This implements a new `_prometheus` endpoint outlined by https://github.com/apache/couchdb/issues/3377

Users can retrieve metrics info via `/_node/{node-name}/_prometheus`. 

**Additional Capability not mentioned in the RFC**
Some users may not want to use authentication
for scraping metrics, so I've added a configurable option that allows scraping via a different port. 

The output looks like this:
```
# TYPE couchdb_couch_log_requests_total counter
couchdb_couch_log_requests_total{level="alert"} 0
couchdb_couch_log_requests_total{level="critical"} 0
couchdb_couch_log_requests_total{level="debug"} 0
couchdb_couch_log_requests_total{level="emergency"} 0
couchdb_couch_log_requests_total{level="error"} 1
couchdb_couch_log_requests_total{level="info"} 5
couchdb_couch_log_requests_total{level="notice"} 10
couchdb_couch_log_requests_total{level="warning"} 0
# TYPE couchdb_couch_replicator_changes_manager_deaths_total counter
couchdb_couch_replicator_changes_manager_deaths_total 0
# TYPE couchdb_couch_replicator_changes_queue_deaths_total counter
couchdb_couch_replicator_changes_queue_deaths_total 0
# TYPE couchdb_couch_replicator_changes_read_failures_total counter
couchdb_couch_replicator_changes_read_failures_total 0
# TYPE couchdb_couch_replicator_changes_reader_deaths_total counter
couchdb_couch_replicator_changes_reader_deaths_total 0
# TYPE couchdb_couch_replicator_checkpoints_failure_total counter
couchdb_couch_replicator_checkpoints_failure_total 0
# TYPE couchdb_couch_replicator_checkpoints_total counter
couchdb_couch_replicator_checkpoints_total 0
# TYPE couchdb_couch_replicator_connection_acquires_total counter
couchdb_couch_replicator_connection_acquires_total 0
# TYPE couchdb_couch_replicator_connection_closes_total counter
couchdb_couch_replicator_connection_closes_total 0
# TYPE couchdb_couch_replicator_connection_creates_total counter
couchdb_couch_replicator_connection_creates_total 0
# TYPE couchdb_couch_replicator_connection_owner_crashes_total counter
couchdb_couch_replicator_connection_owner_crashes_total 0
# TYPE couchdb_couch_replicator_connection_releases_total counter
couchdb_couch_replicator_connection_releases_total 0
# TYPE couchdb_couch_replicator_connection_worker_crashes_total counter
couchdb_couch_replicator_connection_worker_crashes_total 0
# TYPE couchdb_couch_replicator_docs_completed_state_updates_total counter
couchdb_couch_replicator_docs_completed_state_updates_total 0
# TYPE couchdb_couch_replicator_docs_db_changes_total counter
couchdb_couch_replicator_docs_db_changes_total 0
# TYPE couchdb_couch_replicator_docs_dbs_created_total counter
couchdb_couch_replicator_docs_dbs_created_total 0
# TYPE couchdb_couch_replicator_docs_dbs_deleted_total counter
couchdb_couch_replicator_docs_dbs_deleted_total 0
# TYPE couchdb_couch_replicator_docs_failed_state_updates_total counter
couchdb_couch_replicator_docs_failed_state_updates_total 0
# TYPE couchdb_couch_replicator_failed_starts_total counter
couchdb_couch_replicator_failed_starts_total 0
# TYPE couchdb_couch_replicator_jobs_accepting gauge
couchdb_couch_replicator_jobs_accepting 2
# TYPE couchdb_couch_replicator_jobs_accepts_total counter
couchdb_couch_replicator_jobs_accepts_total 16
# TYPE couchdb_couch_replicator_jobs_adds_total counter
couchdb_couch_replicator_jobs_adds_total 0
# TYPE couchdb_couch_replicator_jobs_crashes_total counter
couchdb_couch_replicator_jobs_crashes_total 0
# TYPE couchdb_couch_replicator_jobs_pending gauge
couchdb_couch_replicator_jobs_pending 0
# TYPE couchdb_couch_replicator_jobs_removes_total counter
couchdb_couch_replicator_jobs_removes_total 0
# TYPE couchdb_couch_replicator_jobs_reschedules_total counter
couchdb_couch_replicator_jobs_reschedules_total 2
# TYPE couchdb_couch_replicator_jobs_running gauge
couchdb_couch_replicator_jobs_running 0
# TYPE couchdb_couch_replicator_jobs_starts_total counter
couchdb_couch_replicator_jobs_starts_total 0
# TYPE couchdb_couch_replicator_jobs_stops_total counter
couchdb_couch_replicator_jobs_stops_total 0
# TYPE couchdb_couch_replicator_requests_total counter
couchdb_couch_replicator_requests_total 0
# TYPE couchdb_couch_replicator_responses_failure_total counter
couchdb_couch_replicator_responses_failure_total 0
# TYPE couchdb_couch_replicator_responses_total counter
couchdb_couch_replicator_responses_total 0
# TYPE couchdb_couch_replicator_stream_responses_failure_total counter
couchdb_couch_replicator_stream_responses_failure_total 0
# TYPE couchdb_couch_replicator_stream_responses_total counter
couchdb_couch_replicator_stream_responses_total 0
# TYPE couchdb_couch_replicator_worker_deaths_total counter
couchdb_couch_replicator_worker_deaths_total 0
# TYPE couchdb_couch_replicator_workers_started_total counter
couchdb_couch_replicator_workers_started_total 0
# TYPE couchdb_auth_cache_requests_total counter
couchdb_auth_cache_requests_total 0
# TYPE couchdb_auth_cache_misses_total counter
couchdb_auth_cache_misses_total 0
# TYPE couchdb_collect_results_time_seconds summary
couchdb_collect_results_time_seconds{quantile="0.5"} 0.0
couchdb_collect_results_time_seconds{quantile="0.75"} 0.0
couchdb_collect_results_time_seconds{quantile="0.9"} 0.0
couchdb_collect_results_time_seconds{quantile="0.95"} 0.0
couchdb_collect_results_time_seconds{quantile="0.99"} 0.0
couchdb_collect_results_time_seconds{quantile="0.999"} 0.0
couchdb_collect_results_time_seconds_sum 0.0
couchdb_collect_results_time_seconds_count 0
# TYPE couchdb_couch_server_lru_skip_total counter
couchdb_couch_server_lru_skip_total 0
# TYPE couchdb_database_purges_total counter
couchdb_database_purges_total 0
# TYPE couchdb_database_reads_total counter
couchdb_database_reads_total 0
# TYPE couchdb_database_writes_total counter
couchdb_database_writes_total 0
# TYPE couchdb_db_open_time_seconds summary
couchdb_db_open_time_seconds{quantile="0.5"} 0.0
couchdb_db_open_time_seconds{quantile="0.75"} 0.0
couchdb_db_open_time_seconds{quantile="0.9"} 0.0
couchdb_db_open_time_seconds{quantile="0.95"} 0.0
couchdb_db_open_time_seconds{quantile="0.99"} 0.0
couchdb_db_open_time_seconds{quantile="0.999"} 0.0
couchdb_db_open_time_seconds_sum 0.0
couchdb_db_open_time_seconds_count 0
# TYPE couchdb_dbinfo_seconds summary
couchdb_dbinfo_seconds{quantile="0.5"} 0.0
couchdb_dbinfo_seconds{quantile="0.75"} 0.0
couchdb_dbinfo_seconds{quantile="0.9"} 0.0
couchdb_dbinfo_seconds{quantile="0.95"} 0.0
couchdb_dbinfo_seconds{quantile="0.99"} 0.0
couchdb_dbinfo_seconds{quantile="0.999"} 0.0
couchdb_dbinfo_seconds_sum 0.0
couchdb_dbinfo_seconds_count 0
# TYPE couchdb_document_inserts_total counter
couchdb_document_inserts_total 0
# TYPE couchdb_document_purges_failure_total counter
couchdb_document_purges_failure_total 0
# TYPE couchdb_document_purges_success_total counter
couchdb_document_purges_success_total 0
.
.
.
```

## Testing recommendations
To see metrics flowing, run:

1)`brew install prometheus`
2) Create a prometheus.yml file in a folder with:
```
global:
  scrape_interval:     15s # By default, scrape targets every 15 seconds.

  # Attach these labels to any time series or alerts when communicating with
  # external systems (federation, remote storage, Alertmanager).
  external_labels:
    monitor: 'codelab-monitor'

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # Override the global default and scrape targets from this job every 5 seconds.
    scrape_interval: 30s

    static_configs:
      - targets: ['localhost:15984']
    metrics_path: "/_node/_local/_prometheus"
    basic_auth:
      username: 'adm'
      password: 'pass'
```
3) Launch your cluster with `./dev/run -a adm pass`
4) Launch prometheus with `prometheus --config.file=prometheus.yml` 
5) Open a browser to `http://localhost:9090/metrics` to view metrics data.
For additional info check out https://prometheus.io/docs/introduction/first_steps/

Additional Unit Tests will be added.

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/chttpd/src/chttpd_node.erl [hunk 1]
```diff
@@ -117,6 +117,12 @@ handle_node_req(#httpd{method='GET', path_parts=[_, Node, <<"_stats">> | Path]}=
     chttpd:send_json(Req, EJSON1);
 handle_node_req(#httpd{path_parts=[_, _Node, <<"_stats">>]}=Req) ->
     send_method_not_allowed(Req, "GET");
+handle_node_req(#httpd{method='GET', path_parts=[_, Node, <<"_prometheus">>]}=Req) ->
+    Metrics = call_node(Node, couch_prometheus_server, scrape, []),
+    Header = [{<<"Content-Type">>, <<"text/plain">>}],
```
Comment (rnewson) [inline src/chttpd/src/chttpd_node.erl:122]: I think there's supposed to be a version parameter on this?

File: src/couch_prometheus/src/couch_prometheus_http.erl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                send_resp(MochiReq, 404, [], <<>>)
+        end
+    catch T:R ->
+        Body = io_lib:format("~p:~p", [T, R]),
+        send_resp(MochiReq, 500, [], Body)
+    end.
+
+send_prometheus(MochiReq, Node) ->
+    Headers = couch_httpd:server_header() ++ [
+        {<<"Content-Type">>, <<"text/plain">>}
```
Comment (rnewson) [inline src/couch_prometheus/src/couch_prometheus_http.erl:51]: version param?

File: src/couch_prometheus/src/couch_prometheus_http.erl [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    PathParts =  string:tokens(Path, "/"),    try
+        case PathParts of
+            ["_node", Node, "_prometheus"] ->
+                send_prometheus(MochiReq, Node);
+            _ ->
+                send_resp(MochiReq, 404, [], <<>>)
+        end
+    catch T:R ->
+        Body = io_lib:format("~p:~p", [T, R]),
+        send_resp(MochiReq, 500, [], Body)
```
Comment (rnewson) [inline src/couch_prometheus/src/couch_prometheus_http.erl:46]: we work hard elsewhere not to send ugly errors to the user, can we follow the chttpd:send_error pattern here?

File: src/couch_prometheus/src/couch_prometheus_server.erl [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+-export([
+    start_link/0,
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    code_change/3,
+    terminate/2,
+
+    scrape/0
```
Comment (rnewson) [inline src/couch_prometheus/src/couch_prometheus_server.erl:32]: can we keep the gen_server callbacks in a separate export block to public or private functions pls? we do that ~everywhere else.

File: src/couch_prometheus/src/couch_prometheus_server.erl [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    refresh
+}).
+
+init([]) ->
+    Metrics = refresh_metrics(),
+    RT = update_refresh_timer(),
+    {ok, #st{metrics=Metrics, refresh=RT}}.
+
+scrape() ->
+    {ok, Metrics} = gen_server:call(?MODULE, scrape),
```
Comment (rnewson) [inline src/couch_prometheus/src/couch_prometheus_server.erl:51]: will this ever take >5s?

File: src/couch_prometheus/src/couch_prometheus_sup.erl [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+]).
+
+-define(CHILD(I, Type), {I, {I, start_link, []}, permanent, 5000, Type, [I]}).
+
+start_link() ->
+    supervisor:start_link({local, ?MODULE}, ?MODULE, []).
+
+init([]) ->
+    {ok, {
+        {one_for_one, 5, 10}, [
```
Comment (rnewson) [inline src/couch_prometheus/src/couch_prometheus_sup.erl:29]: how were these numbers chosen?

Comment (rnewson) [reply src/couch_prometheus/src/couch_prometheus_sup.erl:29]: if the couch_prometheus_server process crashing should the couch_prometheus_http be restarted? (i.e, rest_for_one).

**Review by rnewson — COMMENTED** [review]
I like this a lot, I had a few comments.

**Comment by chewbranca** [comment]
Does this need to be a stateful application? Rather than having a dedicated app hierarchy here, can we just add a `_prometheus` endpoint to https://github.com/apache/couchdb/blob/main/src/chttpd/src/chttpd_node.erl#L121-L124 

that then does the conversion over https://github.com/apache/couchdb/blob/main/src/chttpd/src/chttpd_node.erl#L206-L236

We could avoid having the server that does the caching and just do it on demand when the request comes in.

I'm also not a huge fan of introducing an additional http server to the mix. Is supporting unauthenticated requests to the Prometheus endpoint a necessary feature?

File: src/chttpd/src/chttpd_node.erl (hunk 1 above)
Comment (tonysun83) [reply src/chttpd/src/chttpd_node.erl:122]: I don't think there is: https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

File: src/couch_prometheus/src/couch_prometheus_server.erl (hunk 5 above)
Comment (tonysun83) [reply src/couch_prometheus/src/couch_prometheus_server.erl:51]: Only if the message queue is super backed up, but if that's the case, an asynchronous call wouldn't matter either given the queue.

**Comment by tonysun83** [comment]
@chewbranca 

Those were my exact thoughts when I was writing this up as having an app/gen_server was a bit of overkill. The bulk of the code is in the wrapper which does the conversion so it wouldn't make sense to add the extra supervision tree. The two reasons of adding it however were 1) We're going to add _active_task info as well later on and I didn't want to have a huge chunk of code inside chttpd_node to combine 3 different sources of stats info (stats, system, active_tasks) as it'll get messier.  2) The addition of the mochiweb_server was the deciding factor as now we needed a separate process to linger around. Even though it's optional, I'd rather not have a non-supervised mochiweb server lingering in the vm.

The reason for the mochiweb_server was to bypass the need to have a username/password in certain situations. One of those situations that arose was the need to store credentials inside a Prometheus config file. It was an internal scrape but forcing the storage of credentials meant additional security parameters in place such that the credentials would not leak. It would be more secure in that scenario not to use credentials. Thus, it's configurable and OFF by default. Would love some additional thoughts on this cc @janl @wohali.

File: src/couch_prometheus/src/couch_prometheus_sup.erl (hunk 6 above)
Comment (tonysun83) [reply src/couch_prometheus/src/couch_prometheus_sup.erl:29]: The values were copied over directly from couch_stats without much consideration. I think I can make this either configurable or have higher numbers? 

You'e right in that doesn't make sense for  `couch_prometheus_http` to exist without `couch_prometheus_server`, but does restarting really enforce that? I guess if `couch_prometheus_server` keeps crashing till it doesn't crash anymore, then we don't start `couch_prometheus_http`.

**Comment by bessbd** [comment]
@tonysun83 : do we want to have this in `3.x` too?
Other: I couldn't find E2E tests for this PR. Are they in progress or did I miss something?

**Comment by tonysun83** [comment]
@bessbd  I'd like to get some consensus on this before writing E2E tests

**Comment by bessbd** [comment]
> @bessbd I'd like to get some consensus on this before writing E2E tests.

Sure, that works too!
For me personally it's easier to start with the tests when reviewing a PR. That's like a best description of what's happening in a change.

File: src/chttpd/src/chttpd_node.erl (hunk 1 above)
Comment (rnewson) [reply src/chttpd/src/chttpd_node.erl:122]: I'm referring to https://prometheus.io/docs/instrumenting/exposition_formats/

`HTTP Content-Type | text/plain; version=0.0.4 (A missing version value will lead to a fall-back to the most recent text format version.)`

**Review by rnewson — APPROVED** [review]
I think the (admittedly weird) prometheus versioning param should be added but otherwise I'm +1 on this.

File: src/couch_prometheus/src/couch_prometheus_http.erl [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            {ok, Pid};
+        {error, Reason} ->
+            io:format("Failure to start Mochiweb: ~s~n", [Reason]),
+            {error, Reason}
+    end.
+
+handle_request(MochiReq) ->
+    RawUri = MochiReq:get(raw_path),
+    {"/" ++ Path, _, _} = mochiweb_util:urlsplit_path(RawUri),
+    PathParts =  string:tokens(Path, "/"),    try
```
Comment (bessbd) [inline src/couch_prometheus/src/couch_prometheus_http.erl:38]: Is the `try` on the same line on purpose here?

File: src/couch_prometheus/src/couch_prometheus_http.erl [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        case PathParts of
+            ["_node", Node, "_prometheus"] ->
+                send_prometheus(MochiReq, Node);
+            _ ->
+                send_error(MochiReq, 404, <<"not_found">>, <<>>)
+        end
+    catch T:R ->
```
Comment (bessbd) [inline src/couch_prometheus/src/couch_prometheus_http.erl:45]: What kind of errors are possible here?

File: src/couch_prometheus/test/eunit/couch_prometheus_e2e_tests.erl [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+create_db(Url) ->
+    Addr = config:get("chttpd", "bind_address", "127.0.0.1"),
+    Port = mochiweb_socket_server:get(chttpd, port),
+    {ok, Status, _, _} = test_request:put(Url, [?CONTENT_JSON, ?AUTH], "{}"),
+    case Status of
+        201 -> ok;
+        202 -> ok;
+        _ -> io:format(user, "~n HTTP Status Code: ~p~n", [Status])
+    end,
```
Comment (bessbd) [inline src/couch_prometheus/test/eunit/couch_prometheus_e2e_tests.erl:131]: These lines can probably be removed.

**Review by bessbd — APPROVED** [review]
LGTM, +1
I left a couple of nits and one missing license header comment.
Let me know what you think.

Thank you for this work, @tonysun83 !

File: src/couch_prometheus/src/couch_prometheus_http.erl (hunk 8 above)
Comment (tonysun83) [reply src/couch_prometheus/src/couch_prometheus_http.erl:45]: it's just a fail-safe to print out any unknown errors

**Pull request merged**